cloudinit/sources/DataSourceNoCloud.py

```diff
@@ -362,18 +363,50 @@ def _merge_new_seed(cur, seeded):
     return ret
 
 
+def _parse_dmi() -> str:
+    serial = dmi.read_dmi_data("product_serial")
+    if serial:
+        return sources.parse_cmdline_or_dmi(serial)
+    return ""
+
+
 class DataSourceNoCloudNet(DataSourceNoCloud):
     def __init__(self, sys_cfg, distro, paths):
         DataSourceNoCloud.__init__(self, sys_cfg, distro, paths)
         self.supported_seed_starts = ("http://", "https://")
 
     def ds_detect(self):
-        """NoCloud requires "nocloud-net" as the way to specify
-        seeding from an http(s) address. This diverges from all other
-        datasources in that it does a kernel commandline match on something
-        other than the datasource dsname for only DEP_NETWORK.
+        """Check dmi and kernel commandline for dsname
+
+        NoCloud historically used "nocloud-net" as its dsname
+        for network timeframe (DEP_NETWORK), which supports http(s) urls.
+        For backwards compatiblity, check for that dsname.
         """
-        return "nocloud-net" == sources.parse_cmdline()
+        log_deprecated = partial(
+            util.deprecate,
+            deprecated="The 'nocloud-net' datasource name",
+            deprecated_version="24.1",
+            extra_message=(
+                "Use 'nocloud' instead, which uses the seedfrom protocol"
+                "scheme (http// or file://) to decide how to run."
+            ),
+        )
+
+        if "nocloud-net" == sources.parse_cmdline():
```

Comment:
> I think that we also need to change this line to if 'nocloud' in sources.parse_cmdline(), or something similar.

This is [done already in the base class](https://github.com/canonical/cloud-init/blob/beddd97c146e0bdba57b1fadc39e7b707351cef8/cloudinit/sources/__init__.py#L369).

> Otherwise, since nocloud-net is no longer wanted (or documented), specifying nocloud on the kernel CLI with an http(s) server will not be recognized.

We would have [several failing tests](https://github.com/canonical/cloud-init/blob/main/tests/integration_tests/test_kernel_commandline_match.py) if that was broken.